File: update_images.py
import urllib.request
import json
import random

project_id = "opvault-9fc81"
base_url = f"https://firestore.googleapis.com/v1/projects/{project_id}/databases/(default)/documents/cards"

# Fetch all cards
req = urllib.request.Request(base_url)
try:
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode())
except Exception as e:
    print(f"Failed to fetch cards: {e}")
    exit(1)

documents = data.get('documents', [])
if not documents:
    print("No cards found.")
    exit(0)

local_images = [
    "images/card1.png",
    "images/card2.png",
    "images/card3.png",
    "images/card4.png",
    "images/card5.png",
    "images/card6.png",
    "images/card7.png",
    "images/card8.png",
    "images/card9.png"
]

badge_image_map = {}
count = 0

for doc in documents:
    name_path = doc['name']
    fields = doc.get('fields', {})
    
    badge = fields.get('badge', {}).get('stringValue', '')
    
    if badge not in badge_image_map:
        badge_image_map[badge] = random.choice(local_images)
        
    new_image = badge_image_map[badge]
    
    # We need to use the PATCH method to update the image field.
    update_url = f"https://firestore.googleapis.com/v1/{name_path}?updateMask.fieldPaths=image"
    
    patch_data = {
        "fields": {
            "image": {
                "stringValue": new_image
            }
        }
    }
    
    patch_req = urllib.request.Request(update_url, data=json.dumps(patch_data).encode('utf-8'), method='PATCH')
    patch_req.add_header('Content-Type', 'application/json')
    
    try:
        with urllib.request.urlopen(patch_req) as p_res:
            pass
        count += 1
    except Exception as e:
        print(f"Failed to update {name_path}: {e}")

print(f"Successfully updated {count} cards with local images!")

File: update_images.js
const { initializeApp } = require('firebase/app');
const { getFirestore, collection, getDocs, doc, updateDoc } = require('firebase/firestore');

const firebaseConfig = {
    apiKey: "AIzaSyB8uFC6UUXnO3esLIB1xor51fqy0spi1ZY",
    authDomain: "opvault-9fc81.firebaseapp.com",
    projectId: "opvault-9fc81",
    storageBucket: "opvault-9fc81.firebasestorage.app",
    messagingSenderId: "207978482582",
    appId: "1:207978482582:web:41022514fd338a6168abb7",
    measurementId: "G-TMN4T4L7XG"
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

const localImages = [
    "images/card1.png",
    "images/card2.png",
    "images/card3.png",
    "images/card4.png",
    "images/card5.png",
    "images/card6.png",
    "images/card7.png",
    "images/card8.png",
    "images/card9.png"
];

function randomImage() {
    return localImages[Math.floor(Math.random() * localImages.length)];
}

async function updateImages() {
    const querySnapshot = await getDocs(collection(db, "cards"));
    
    // Map to keep track of which badge gets which image
    const badgeImageMap = {};
    let promises = [];
    let count = 0;

    querySnapshot.forEach((document) => {
        const data = document.data();
        const badge = data.badge; // Using badge as the unique identifier for a card type
        
        if (!badgeImageMap[badge]) {
            badgeImageMap[badge] = randomImage();
        }
        
        const newImage = badgeImageMap[badge];
        
        promises.push(
            updateDoc(doc(db, "cards", document.id), {
                image: newImage
            })
        );
        count++;
    });

    await Promise.all(promises);
    console.log(`Successfully updated ${count} cards with local images!`);
    process.exit(0);
}

updateImages().catch(console.error);
